test/commands/scanner/rule/list.test.ts

```diff
@@ -21,91 +21,127 @@ function listContentsOverlap<T>(list1: T[], list2: T[]): boolean {
 	return list1.some(x => list2.includes(x));
 }
 
+/**
+ * Rather than painstakingly check all of the rules, we'll just make sure that we got the right number of rules,
+ * compared to the number of rules in the catalog. This method filters the catalogs rules.
+ *
+ * @param includeDefaultDisabled - if true, include rule when Rule.defaultEnabled is false. if false, include rule when Rule.defaultEnabled is true
+ * @param includedCategories - include rule if any of its categories overlaps with this array
+ * @param excludeCategories- exclude rule if any of its categories overlaps with this array
+ *
+ * @return the number of rules returned
+ */
+async function getRulesFilteredByCategoryCount (includeDefaultDisabled: boolean, includedCategories: string[]=undefined, excludeCategories: string[]=undefined): Promise<number> {
+	const catalog = getCatalogJson();
+	const enabledEngineNames = (await Controller.getEnabledEngines()).map(e => e.getName());
+	let rules: Rule[] = catalog.rules.filter(r => (includeDefaultDisabled || r.defaultEnabled) && enabledEngineNames.includes(r.engine));
+	if (includedCategories?.length > 0) {
```

Comment:
Curious what people think of this syntax versus (includedCategories && includedCategories.length > 0)

Reply:
Since `includedCategories` and `excludedCategories` are never simultaneously non-empty, could you combine them into a single parameter and supply a `isNegated` boolean? That should let you consolidate the two `if` branches into a single branch.

Reply:
I gave this a try, it seems more complex. The negated and categories are defaulted, but it would be likely be a developer error to only pass in one of the values. Thoughts?

```Typescript
async function getRulesFilteredByCategoryCount(includeDefaultDisabled: boolean, negated=false, categories: string[]=undefined): Promise<number> {
	const catalog = getCatalogJson();
	const enabledEngineNames = (await Controller.getEnabledEngines()).map(e => e.getName());
	let rules: Rule[] = catalog.rules.filter(r => (includeDefaultDisabled || r.defaultEnabled) && enabledEngineNames.includes(r.engine));
	if (categories?.length > 0) {
		rules = rules.filter((r: Rule) => negated ? !categories.some(c => r.categories.includes(c)) : categories.some(c => r.categories.includes(c)));
	}
	expect(rules.length).to.be.above(0, 'Expected rule count cannot be zero. Test invalid');
	return rules.length;
}
```

Reply:
Perhaps:
```
if (categories?.length > 0) {
    let categoryPresent: (r: Rule) => boolean = (r: Rule) => categories.some(c => r.categories.includes(c));
    rules = rules.filter((r: Rule) => negated ? !categoryPresent(r) : categoryPresent(r));
}
```

Reply:
While more verbose, I think the current code is clearer. I'd like to go with the code as-is for now, we can discuss more at standup if there is a strong desire to change.